src/cart/crt_hlc.c

```diff
@@ -65,11 +65,14 @@ uint64_t crt_hlc_get(void)
 	return ret;
 }
 
-uint64_t crt_hlc_get_msg(uint64_t msg)
+int crt_hlc_get_msg(uint64_t msg, uint64_t *hlc_out)
 {
 	uint64_t pt = crt_hlc_localtime_get();
 	uint64_t hlc, ret, ml = msg & ~CRT_HLC_MASK;
 
+	if (ml > pt && ml - pt > crt_hlc_epsilon)
+		return -DER_HLC_SYNC;
```

Comment:
How is this error supposed to be handled? Does it indicate a 'fatal' condition or is it something we expect to happen intermittently with rpc being retried upon this error?
Also, why is swim being isolated from those errors, and not other internal rpcs, for example uri lookup?

Reply:
> How is this error supposed to be handled? Does it indicate a 'fatal' condition or is it something we expect to happen intermittently with rpc being retried upon this error?

It indicates a fatal condition where servers' clocks are not properly synchronized (via NTP, for instance). Under such conditions, a DAOS transaction may not "see" other conflicting transactions committed before it starts. E.g., a DFS open may not "see" the file that has already been created recently. The handling involves two parts:

1. DAOS may increase epsilon to cope with the unexpected clock offsets. This isn't available at the moment.
2. The administrator needs to fix the clock synchronization setup.

> Also, why is swim being isolated from those errors, and not other internal rpcs, for example uri lookup?

It's unclear what would happen if we were to reject SWIM messages here. I think it might cause an innocent server (i.e., who's clock is correct) to be evicted due to a different, faulty server. As to other RPCs, we could use an RPC registration flag to indicate if an RPC operation relies on HLC timestamp ordering or not. If it does not, then we don't have to reject it.

Reply:
An easier alternative to the flag is maybe allowing all crt internal RPCs except for IV ones. crt must guarantee that it will never rely on HLC timestamp ordering.